cmake: fix relative path calculate error

LingaoM opened the pull request:
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/74647

<img width="1210" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/26378606/98f7fdbd-6365-4fb4-9554-3c6b43766f5d">

ithinuel reviewed (changes requested):
This removes the use of relative path which was intentionally added in https://github.com/zephyrproject-rtos/zephyr/pull/16368.

cc @marc-hb is it still required ?

marc-hb commented:
> The build directory absolutely is local generated directory, not public for other.

No, the build directory is not "100% private": to make sure the build is reproducible you need to compare object files.

Reproducible builds are important for security, for instance they help with "JiaT75" issues. Comparing build directories also helps catching small but harmful differences between CI and developers and between developers.
 
For more context see #50205, #14593, https://www.cisa.gov/sbom, https://reproducible-builds.org/ etc.

marc-hb reviewed (dismissed):
> This removes the use of relative path which was intentionally added in https://github.com/zephyrproject-rtos/zephyr/pull/16368.

This is indeed a logical revert of d1b4da96e1484a276c21c3, so at least the commit message is wrong. It currently says "fix relative path calculate error" but this commit does not fix any "calculate error", instead it reverts to an absolute, non reproducible path.

Thanks @ithinuel for the good catch.

LingaoM committed:
cmake: fix relative path calculate error

```C
file(RELATIVE_PATH relpath ${ZEPHYR_BASE}/include ${path})
```

The prerequisite path assumed here must be the prefix of ZEPHYR,
but if use symbolic links it will calculate error: like this:
https://github.com/zephyrproject-rtos/zephyr/issues/74647#issuecomment-2183912468

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/74647

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
@marc-hb @ithinuel I've reworked a version that will still use relative paths, but adds support for link paths.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -1335,7 +1335,10 @@ function(zephyr_linker_sources location)
 
     # Find the relative path to the linker file from the include folder.
-    file(RELATIVE_PATH relpath ${ZEPHYR_BASE}/include ${path})
+    get_filename_component(ZEPHYR_BASE_REALPATH "${ZEPHYR_BASE}/include" REALPATH)
+    get_filename_component(PATH_REALPATH "${path}" REALPATH)
```
The command `get_filename_component(... REALPATH)` has been superseded by `file(REAL_PATH ...)` cmake.
> `get_filename_component`
> 
> Get a specific component of a full filename.
>
> Changed in version 3.20: This command has been superseded by the [cmake_path()](https://cmake.org/cmake/help/latest/command/cmake_path.html#command:cmake_path) command, except for REALPATH, which is now offered by [file(REAL_PATH)](https://cmake.org/cmake/help/latest/command/file.html#real-path),

https://cmake.org/cmake/help/latest/command/get_filename_component.html#command:cmake_path

but please note that `file(REAL_PATH ...)` is affected by [CMP0152](https://cmake.org/cmake/help/latest/policy/CMP0152.html)., and before you ask, then yes, `get_filename_component(... REAL_PATH)` suffers the same issue related to symlink resolving, but doesn't have a policy for controlling this (always fixed at OLD behavior).

Which again shows that the use of symlinks can be tricky.

Also, variables intended for local use should be written in lowercase, like `path_realpath` and `zephyr_base_realpath`.
And please use another name than `zephyr_base_realpath`, as it is not the real path of Zephyr you are defining, but a subfolder (`include`).
